File: Python/MongoDB/CRUD/mongodb_crud_client.py
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi
from bson.objectid import ObjectId

class MongoDB_CRUD_Client:
    def __init__(self, config):
        self.config = config
        self.uri = f"mongodb+srv://{config['username']}:{config['password']}@{config['server']}/?retryWrites={config['retryWrites']}&w={config['w']}&appName={config['appName']}"
        self.client = None
    
    def __get_client(self):
        if self.client is None:
            self.client = MongoClient(self.uri, server_api=ServerApi('1'))
        return self.client

    def ping(self):
        self.__get_client().admin.command('ping')

    def close(self):
        if self.client is not None:
            self.client.close()
            self.client = None

    def get_collection(self, db_name, collection_name):
        client = self.__get_client()
        return client[db_name][collection_name]
    
    def create(self, db_name, collection_name, document) -> ObjectId:
        collection = self.get_collection(db_name, collection_name)
        result = collection.insert_one(document)
        return result.inserted_id

    def read(self, db_name, collection_name, document_id: ObjectId):
        collection = self.get_collection(db_name, collection_name)
        document = collection.find_one({"_id": document_id})
        return document

    def update(self, db_name, collection_name, document_id: ObjectId, update_fields):
        collection = self.get_collection(db_name, collection_name)
        result = collection.find_one_and_update({"_id": document_id}, {"$set": update_fields}, new=True)
        return result

    def delete(self, db_name, collection_name, document_id: ObjectId):
        collection = self.get_collection(db_name, collection_name)
        result = collection.delete_one({"_id": document_id})
        return result.deleted_count == 1

File: JavaScript/MongoDB/CRUD/MongoDB_CRUD_Client.js
import { MongoClient, ServerApiVersion } from 'mongodb';

class MongoDB_CRUD_Client {
    constructor(config) {
        this.config = config;
        this.uri = `mongodb+srv://${config.username}:${config.password}@${config.server}/?retryWrites=${config.retryWrites}&w=${config.w}&appName=${config.appName}`;
        this.client = null;
    }

    async __get_client() {
        if (this.client === null) {
            this.client = new MongoClient(this.uri, { serverApi: ServerApiVersion.v1 });
            await this.client.connect();
        }
        return this.client;
    }

    async get_collection(db_name, collection_name) {
        const client = await this.__get_client();
        return client.db(db_name).collection(collection_name);
    }

    async create(db_name, collection_name, document) {
        const collection = await this.get_collection(db_name, collection_name);
        const result = await collection.insertOne(document);
        return result.insertedId;
    }

    async read(db_name, collection_name, document_id) {
        const collection = await this.get_collection(db_name, collection_name);
        const document = await collection.findOne({ _id: document_id });
        return document;
    }

    async update(db_name, collection_name, document_id, update_fields) {
        const collection = await this.get_collection(db_name, collection_name);
        const result = await collection.findOneAndUpdate(
            { _id: document_id },
            { $set: update_fields },
            { returnDocument: 'after' }
        );
        return result.value;
    }

    async delete(db_name, collection_name, document_id) {
        const collection = await this.get_collection(db_name, collection_name);
        const result = await collection.deleteOne({ _id: document_id });
        return result.deletedCount === 1;
    }

    async ping() {
        const client = await this.__get_client();
        await client.db('admin').command({ ping: 1 });
    }

    async close() {
        if (this.client !== null) {
            await this.client.close();
            this.client = null;
        }
    }
}

export default MongoDB_CRUD_Client;